### DISCUSSION TITLE
"Interesting" (or not) PHP switch processing in PHP 8.0 and later

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#6083] [General]
Let's say you have a PHP switch statement like the following:
```php
        switch ($result) {
            case -1:
                $error = 'Value is -1';
                break;
            case -2:
            case -3:
            case -4:
                $error = "The result is $result";
                break;
            case false:
                $error = 'The result is false';
                break;
            default:
                $error = '';
                break;
        }
        if ($error !== '') {
            echo $error;
        } else {
            echo 'No error!';
        }
```

Further, let's pretend that the `$result` is `(bool)true`; what's the expected output?  The results will surprise you (they did me)!

**Comment by lat9** [comment]
Hint: It's got something to do with this PHP RFC: https://wiki.php.net/rfc/string_to_number_comparison

**Reply by torvista** [reply]
Reinforces my custom to declare strict all the files I touch....perhaps it should be policy on modified files as we progress....

**Reply by drbyte** [reply]
You may be similarly surprised to learn that even declaring strict_types on lat9's example doesn't change/fix the processing.

Nevertheless, while setting it to `true` and having it say `Value is -1`, you probably wouldn't have then expected that if you comment-out the `-1` case and let it go to -2/-3/-4 ... it comes back saying `The result is 1`.

**Reply by torvista** [reply]
I meant that setting strict types seems to be helpful/useful to catch low-hanging type mis-use.

**Reply by lat9** [reply]
Yep, I corrected the switch by casting the `$result` to an int and changing the `false` to 0.  I lost 3 _frustrating_ hours of my life last night trying to figure out what the heck was going on.

**Comment by drbyte** [comment]
One way I've often seen done is a slightly different syntax:

```php
switch (true) {
    case ($result === true): // put the type-specific checking in the `case` statement. Parentheses are technically optional, depending on complexity.
    ...
//etc
}

**Reply by lat9** [reply]
FWIW, that `select` clause (albeit somewhat modified) was pulled from `paypaldp.php`.
